src/main/java/graphql/nadel/engine/SourceQueryTransformer.java

```diff
@@ -0,0 +1,124 @@
+package graphql.nadel.engine;
+
+import graphql.execution.ExecutionContext;
+import graphql.execution.ExecutionStepInfo;
+import graphql.execution.ExecutionStepInfoFactory;
+import graphql.execution.MergedField;
+import graphql.language.Document;
+import graphql.language.Field;
+import graphql.language.FieldDefinition;
+import graphql.language.OperationDefinition;
+import graphql.language.Selection;
+import graphql.language.SelectionSet;
+import graphql.nadel.dsl.FieldDefinitionWithTransformation;
+import graphql.nadel.dsl.FieldTransformation;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static graphql.execution.MergedField.newMergedField;
+import static graphql.language.OperationDefinition.newOperationDefinition;
+import static graphql.language.SelectionSet.newSelectionSet;
+import static java.util.Collections.singletonList;
+
+public class SourceQueryTransformer {
```

Comment:
this whole class should take advantage of `AstTransformer` or maybe `QueryVisitor`: both traverse the query for you. Especially `QueryVisitor` follows Fragments.

Reply:
We need to also add unit tests at some point. Could be later if we decide so.

Reply:
yup, unit testing is very important here, just wanted to get initial feedback